### PR TITLE
Add validation for snapshot consistency

### PR DESCRIPTION
**Pull request by madmax983**
Added validation logic to `tests/snapshot_race_condition.rs` to detect inconsistent snapshots.
- Updated `HistoricalStorage` usage to allow concurrent writes via `RwLock`.
- Modified test writer thread to update both `current` and `historical` storage, simulating a real transaction.
- Implemented checks for "orphaned versions" (historical version exists but node missing in current) and "missing history" (node exists in current but version missing in historical).
- Verified that the test fails when enabled (confirming it catches the race condition) and is properly ignored for now.

---
*PR created automatically by Jules for task [17207235228954240963](https://jules.google.com/task/17207235228954240963) started by @madmax983*

### CODE REVIEW

File: tests/snapshot_race_condition.rs [hunk 1]
```diff
@@ -63,19 +78,28 @@ fn test_concurrent_write_during_snapshot_creation() {
...
             // It should NEVER happen between the two snapshot creations!
+
+            // Update both storages (simulating transaction)
             let _ = current.insert_node_direct(node, time::now());
+            // In a real DB, these happen together. Here we do them sequentially which
+            // increases the chance of race if checkpoint interleaves.
+            let _ = historical
+                .write()
+                .unwrap()
+                .add_node_version(node_id, version_id, temporal, label, props);
```
[the remark this answers is not in the document]
Comment (madmax983) [reply tests/snapshot_race_condition.rs]: > medium
> 
> The `Result`s from `insert_node_direct` and `add_node_version` are currently ignored. In a test, it's better to handle these results explicitly by calling `.unwrap()`. This ensures that any unexpected write failure will cause the test to panic immediately, making it easier to diagnose issues, rather than letting the test fail later with a less obvious error.
> 
> For example, line 92 could be changed to:
> 
> ```rust
> current.insert_node_direct(node, time::now()).unwrap();
> ```
> 
> And lines 95-98 could be changed to:
> 
> ```rust
> historical
>     .write()
>     .unwrap()
>     .add_node_version(node_id, version_id, temporal, label, props)
>     .unwrap();
> ```

@jules

File: tests/snapshot_race_condition.rs [hunk 2]
```diff
@@ -92,8 +116,51 @@ fn test_concurrent_write_during_snapshot_creation() {
...
+    for version in recovered_historical.__test_get_node_versions_iterator() {
+        if version.temporal.valid_time().is_current() {
+            if recovered_current.get_node(version.node_id).is_err() {
+                panic!(
+                    "Found orphaned version: {:?} for node {:?} (exists in historical but not current)",
+                    version.id, version.node_id
+                );
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (madmax983) [reply tests/snapshot_race_condition.rs]: > medium
> 
> The loop for checking orphaned versions can be expressed more idiomatically and concisely using the `find` iterator combinator. This makes the intent of finding the first orphaned version clearer and is a common pattern in Rust for such checks.
> 
> ```rust
>     if let Some(version) = recovered_historical
>         .__test_get_node_versions_iterator()
>         .find(|v| v.temporal.valid_time().is_current() && recovered_current.get_node(v.node_id).is_err())
>     {
>         panic!(
>             "Found orphaned version: {:?} for node {:?} (exists in historical but not current)",
>             version.id, version.node_id
>         );
>     }
> ```

@jules

File: tests/snapshot_race_condition.rs [hunk 3]
```diff
@@ -92,8 +116,51 @@ fn test_concurrent_write_during_snapshot_creation() {
...
+        if recovered_historical
+            .get_current_node_version(node_id)
+            .is_none()
+        {
+            panic!(
+                "Found node without history: {:?} (exists in current but no current version in historical)",
+                node_id
+            );
+        }
+    }
```
[the remark this answers is not in the document]
Comment (madmax983) [reply tests/snapshot_race_condition.rs:187]: > medium
> 
> Similar to the "Orphaned Versions" check, this loop for finding nodes with missing history can be refactored using `find` on an iterator. This approach is more functional and clearly expresses the goal of finding the first occurrence of the inconsistency.
> 
> ```rust
>     if let Some(node_id) =
>         recovered_current.get_all_node_ids().into_iter().find(|&node_id| {
>             recovered_historical
>                 .get_current_node_version(node_id)
>                 .is_none()
>         })
>     {
>         panic!(
>             "Found node without history: {:?} (exists in current but no current version in historical)",
>             node_id
>         );
>     }
> ```

@jules

**Comment by madmax983** [comment]
> ## Code Review: Add validation for snapshot consistency
> ### Overview
> This PR adds important validation logic to detect inconsistent snapshots caused by race conditions during checkpoint creation. The test simulates concurrent writes while creating checkpoints and validates that the recovered state is consistent.
> 
> ### ✅ **Strengths**
> 1. **Addresses Real Concurrency Issue**: The PR correctly identifies and tests for race conditions between checkpoint creation and concurrent writes.
> 2. **Comprehensive Validation Logic**: The two validation checks are well-designed:
>    
>    * **Orphaned versions**: Detects historical versions without corresponding current nodes
>    * **Missing history**: Detects current nodes without historical versions
> 3. **Realistic Test Scenario**: Simulating writes to both `current` and `historical` storage mirrors real transaction behavior.
> 4. **Proper Resource Cleanup**: Uses `tempdir().keep()` to persist for recovery and cleans up at the end.
> 
> ### ⚠️ **Issues & Concerns**
> #### **1. Critical: Incorrect RwLock Usage in Checkpoint Thread**
> **Location**: Lines 49-67
> 
> ```rust
> let historical_guard = historical_clone.read().unwrap();
> let stats = manager
>     .create_checkpoint(LSN(1), &current_clone, &*historical_guard)
>     .unwrap();
> ```
> 
> **Problem**: The checkpoint thread holds a **read lock** (`.read()`) on `HistoricalStorage`, but the writer thread also needs to acquire a **write lock** (`.write()`) to add versions. This creates a **deadlock scenario** or **lock contention** that defeats the purpose of testing the race condition.
> 
> **Expected Behavior**: The test should allow true concurrent access. The lock should either:
> 
> * Be held for the entire checkpoint duration (blocking writes - which is the correct fix for the race)
> * Or not be held at all (demonstrating the race condition exists)
> 
> **Current State**: The read lock allows multiple readers but blocks writers for the entire checkpoint duration, which changes the timing characteristics you're trying to test.
> 
> #### **2. Error Handling: Unwrap Usage**
> **Location**: Multiple locations (lines 32, 40, 59, 66, 96, etc.)
> 
> ```rust
> .unwrap()
> ```
> 
> **Problem**: Per [CODING_STANDARDS.md](../docs/CODING_STANDARDS.md), production code should never use `.unwrap()` or `.expect()`. While this is test code, using `unwrap()` hides the actual error that occurred, making debugging harder.
> 
> **Recommendation**: Use `?" or provide descriptive error messages:
> 
> ```rust
> .expect("Failed to add node version during test setup")
> ```
> 
> #### **3. Race Condition Still Present**
> **Location**: Lines 92-97
> 
> ```rust
> // Update both storages (simulating transaction)
> let _ = current.insert_node_direct(node, time::now());
> // In a real DB, these happen together. Here we do them sequentially which
> // increases the chance of race if checkpoint interleaves.
> let _ = historical
>     .write()
>     .unwrap()
>     .add_node_version(node_id, version_id, temporal, label, props);
> ```
> 
> **Problem**: The comment correctly identifies that these two operations happen sequentially, not atomically. However, there's a **second race** not mentioned: the writer obtains the `.write()` lock **separately** from the insert to `current`. This means:
> 
> 1. Writer inserts to `current`
> 2. Checkpoint reads `current` (sees new node)
> 3. Writer acquires historical lock and adds version
> 4. Checkpoint reads `historical` (may or may not see new version)
> 
> This is actually **correct for testing** - you want to reproduce the race. But the test currently **won't reliably fail** because:
> 
> * The read lock held by the checkpoint thread will block the writer's `.write()` call
> * This reduces the race condition window significantly
> 
> #### **4. Missing Thread Synchronization Gap**
> **Problem**: The `barrier.wait()` ensures both threads start simultaneously, but there's no mechanism to ensure the writer actually executes **during** checkpoint creation. The writer might finish before the checkpoint even starts, or start after it's done.
> 
> **Recommendation**: Add a small delay or use more sophisticated synchronization:
> 
> ```rust
> barrier_clone.wait();
> std::thread::sleep(std::time::Duration::from_micros(100)); // Let checkpoint start
> ```
> 
> #### **5. Test Configuration: Ignored Test**
> **Location**: Line 21 (`#[ignore]`)
> 
> **Problem**: The test is ignored, which means it won't run in CI/CD. While the PR description mentions it "fails when enabled", having a test that's always ignored reduces its value.
> 
> **Recommendation**: Consider one of these approaches:
> 
> 1. **Expected failure test**: Use `#[should_panic]` if you expect it to fail
> 2. **Feature flag**: Gate behind a feature flag for stress testing
> 3. **Fix and enable**: If this detects a real bug, file an issue and enable the test once fixed
> 
> #### **6. Resource Leak: WAL Directory**
> **Location**: Lines 123-124
> 
> ```rust
> let wal_dir = data_path.join("wal_recovery");
> std::fs::create_dir_all(&wal_dir).unwrap();
> ```
> 
> **Problem**: The WAL directory is created but never explicitly cleaned up. While `data_path` is removed at line 162, it's good practice to be explicit.
> 
> **Note**: This is minor since the parent directory cleanup will handle it.
> 
> ### 🔍 **Design Considerations**
> #### **RwLock vs Mutex**
> The PR changes `HistoricalStorage` from `Arc` to `Arc<RwLock<>>`. This is a significant API change that affects all callers.
> 
> **Questions**:
> 
> 1. Does `HistoricalStorage` have internal synchronization? (Looking at the code, it likely doesn't for performance)
> 2. Is `RwLock` the right choice, or should `HistoricalStorage` use internal locks for specific operations?
> 3. Should checkpoint creation take exclusive locks on both storages?
> 
> **Architectural Implication**: The real fix is likely to make checkpoint creation **atomic** by holding write locks on both `current` and `historical` for the entire snapshot duration. This test should verify that behavior.
> 
> #### **Missing Test Cases**
> Consider adding tests for:
> 
> 1. **Write before checkpoint**: Verify it appears in both snapshots
> 2. **Write after checkpoint**: Verify it appears in neither snapshot
> 3. **Multiple concurrent writers**: Stress test with N writers
> 4. **Partial write failure**: What happens if one storage write succeeds but the other fails?
> 
> ### 📊 **Performance Impact**
> Adding `RwLock` to `HistoricalStorage` will impact all historical queries. Even read-only operations now require lock acquisition.
> 
> **Recommendation**: Benchmark historical query performance before/after this change.
> 
> ### 🔒 **Security Considerations**
> No security issues identified. The test uses validated `NodeId::new()`, following the ID security guidelines from CODING_STANDARDS.md.
> 
> ### ✅ **Test Coverage**
> The validation logic (lines 118-158) is thorough and tests the right invariants. Good work on:
> 
> * Checking both directions (orphaned versions + missing history)
> * Using proper error messages with `panic\!`
> * Cleaning up resources
> 
> ### 📝 **Recommendations**
> #### **High Priority**
> 1. **Fix RwLock usage**: The checkpoint thread should either:
>    
>    * Not hold locks at all (to reproduce the race)
>    * Hold write locks (to demonstrate the fix)
> 2. **Add explanatory comments**: Document WHY the test is ignored and what fix is needed
> 3. **Consider test structure**: Should this be:
>    
>    * One test for "demonstrates the bug" (currently panics, ignored)
>    * One test for "verifies the fix" (currently passing, enabled)
> 
> #### **Medium Priority**
> 4. **Replace unwrap() calls**: Use `?` or `.expect()` with descriptive messages
> 5. **Add timing verification**: Ensure writer executes during checkpoint
> 6. **Add performance benchmarks**: If `RwLock` becomes part of the main codebase
> 
> #### **Low Priority**
> 7. **Extract validation logic**: The recovery validation (lines 118-158) could be a reusable test utility
> 8. **Add more test scenarios**: Multiple writers, write timing variations
> 
> ### 🎯 **Suggested Changes**
> ```rust
> // Option 1: Demonstrate the race (no locks held during checkpoint)
> let checkpoint_thread = thread::spawn(move || {
>     barrier_clone.wait();
>     let dir = tempdir().unwrap();
>     let path = dir.keep();
>     let config = CheckpointConfig::with_data_dir(&path);
>     let mut manager = CheckpointManager::new(config).unwrap();
>     
>     // Don't hold any locks - let the race happen
>     let historical_snapshot = historical_clone.read().unwrap();
>     let stats = manager
>         .create_checkpoint(LSN(1), &current_clone, &*historical_snapshot)
>         .unwrap();
>     drop(historical_snapshot); // Release immediately
>     
>     (stats, path)
> });
> 
> // Option 2: Verify the fix (hold write locks for atomicity)
> let checkpoint_thread = thread::spawn(move || {
>     barrier_clone.wait();
>     
>     // Acquire write lock to block all writers during checkpoint
>     let mut historical_guard = historical_clone.write().unwrap();
>     
>     let dir = tempdir().unwrap();
>     let path = dir.keep();
>     let config = CheckpointConfig::with_data_dir(&path);
>     let mut manager = CheckpointManager::new(config).unwrap();
>     
>     let stats = manager
>         .create_checkpoint(LSN(1), &current_clone, &mut *historical_guard)
>         .unwrap();
>     
>     (stats, path)
> });
> ```
> 
> ### ✅ **Conclusion**
> This PR adds valuable test coverage for a critical concurrency issue. The validation logic is well-designed and comprehensive. However, the `RwLock` usage needs refinement to actually reproduce (or prevent) the race condition being tested.
> 
> The test infrastructure is solid - once the locking strategy is clarified, this will be an excellent regression test.
> 
> **Recommendation**: **Request changes** to clarify the locking strategy and fix the test to either consistently reproduce the race or verify the fix.

@jules

**Comment by madmax983** [comment]
> # Pull Request Review: Add validation for snapshot consistency
> ## Summary
> This PR adds validation logic to detect inconsistencies between current and historical storage snapshots during checkpoint creation. The test successfully demonstrates the race condition and provides proper validation checks.
> 
> ## Strengths ✅
> ### 1. **Excellent Test Design**
> * Clear reproduction of the race condition using barriers and concurrent threads
> * Well-documented comments explaining the "torn read" scenario (lines 61-70)
> * Two-level validation: both orphaned versions AND missing history checks
> 
> ### 2. **Good Architecture Understanding**
> * Properly wraps `HistoricalStorage` in `RwLock` to enable concurrent access patterns
> * Correctly initializes both current and historical storage together (lines 37-42)
> * Uses `tempdir().keep()` to persist checkpoint data for inspection
> 
> ### 3. **Comprehensive Validation**
> * Check 1 (lines 136-148): Detects orphaned versions in historical storage
> * Check 2 (lines 150-162): Detects missing history for current nodes
> * Proper cleanup with `std::fs::remove_dir_all()`
> 
> ## Issues to Address 🔧
> ### 1. **Error Handling: Multiple `unwrap()` calls**
> **Severity: Medium**
> 
> While `unwrap()` is acceptable in tests per [CODING_STANDARDS.md](https://github.com/madmax983/GallifreyDB/blob/trunk/docs/CODING_STANDARDS.md#never-use-unwrapexpect-in-production), the test should still handle potential failures gracefully to provide better diagnostics.
> 
> **Problematic areas:**
> 
> * Line 57: `dir.keep()` - should handle potential error with `expect()`
> * Line 62: `historical_guard = historical_clone.read().unwrap()` - poison lock handling
> * Line 103-104: Thread join `unwrap()` - should propagate panic information
> 
> **Recommendation:** Replace key `unwrap()` calls with `expect()` and descriptive messages:
> 
> ```rust
> let path = dir.keep().expect("Failed to persist temp directory");
> let historical_guard = historical_clone.read()
>     .expect("Failed to acquire read lock on historical storage");
> ```
> 
> ### 2. **Missing Sleep Justification**
> **Severity: Low**
> 
> Line 74-76 has a comment about "small delay" but it's not in the actual code shown in the diff. The diff shows:
> 
> ```rust
> std::thread::sleep(std::time::Duration::from_millis(10));
> ```
> 
> This sleep is a timing-based test antipattern. While it increases reproducibility, it:
> 
> * Makes tests slower (10ms per run)
> * Doesn't guarantee race condition will occur
> * Is non-deterministic
> 
> **Recommendation:** Add a comment explaining this is intentional for race reproduction:
> 
> ```rust
> // Intentional delay to increase probability of checkpoint thread
> // acquiring lock before writer, creating the "torn read" condition
> std::thread::sleep(std::time::Duration::from_millis(10));
> ```
> 
> ### 3. **Resource Cleanup Could Be More Robust**
> **Severity: Low**
> 
> Line 165: `std::fs::remove_dir_all(data_path).unwrap()` will panic if cleanup fails, potentially masking test failures.
> 
> **Recommendation:** Use a cleanup guard or handle the error gracefully:
> 
> ```rust
> // Cleanup - ignore errors as test has already passed/failed
> let _ = std::fs::remove_dir_all(data_path);
> ```
> 
> ### 4. **Missing Test for the Passing Case**
> **Severity: Low**
> 
> The test is currently `#[ignore]` because it's expected to fail. However, there's no complementary test that verifies the fix works once implemented.
> 
> **Recommendation:** Add a comment or tracking issue reference:
> 
> ```rust
> #[test]
> #[ignore] // Will fail until fix is implemented (Issue #XXX)
> fn test_concurrent_write_during_snapshot_creation() {
> ```
> 
> ## Performance Considerations 🚀
> * The test creates 200 nodes with full historical versions - reasonable for a race condition test
> * WAL system creation (line 131) is heavyweight but necessary for recovery testing
> * Overall test complexity is appropriate for what it's validating
> 
> ## Security Considerations 🔒
> * No security concerns - this is test code only
> * Proper ID validation via `NodeId::new()` (lines 32, 79)
> * No unsafe code or FFI boundaries
> 
> ## Test Coverage 📊
> This test adds valuable coverage for:
> 
> * Concurrent checkpoint creation scenarios
> * Snapshot consistency validation
> * Recovery from checkpoint with validation
> 
> **Recommendation:** Ensure this test is run as part of CI once the fix is implemented (remove `#[ignore]`).
> 
> ## Code Quality Metrics ⭐
> * **Readability:** Excellent - clear comments and structure
> * **Maintainability:** Good - well-organized with clear validation logic
> * **Follows Standards:** Mostly yes - minor improvements needed for error handling
> 
> ## Action Items
> Before merging:
> 
> * [ ]  Replace critical `unwrap()` calls with `expect()` and descriptive messages
> * [ ]  Add comment explaining intentional sleep delay
> * [ ]  Make resource cleanup more robust (ignore cleanup errors)
> * [ ]  Add issue reference to `#[ignore]` attribute
> * [ ]  Verify `cargo fmt` has been run
> * [ ]  Verify `cargo clippy --all-targets --all-features -- -D warnings` passes
> 
> ## Overall Assessment
> **Status:** Approve with minor changes ✅
> 
> This is a well-designed TDD test that properly demonstrates the race condition issue. The validation logic is comprehensive and will catch the bug once the test is enabled. The suggested improvements are minor and mostly about defensive error handling and documentation.
> 
> Great work on creating a reproducible test case for this complex concurrency issue!
> 
> _Review conducted according to [CLAUDE.md](https://github.com/madmax983/GallifreyDB/blob/trunk/CLAUDE.md) and [CODING_STANDARDS.md](https://github.com/madmax983/GallifreyDB/blob/trunk/docs/CODING_STANDARDS.md)_

@jules

**Comment by madmax983** [comment]
> ## Code Review for PR #602: Add validation for snapshot consistency
> ### Summary
> This PR adds important validation logic to detect inconsistent snapshots in concurrent write scenarios. Overall well-thought-out, but several issues need addressing.
> 
> ## Critical Issues
> **1. Resource Leak: Tempdir Not Properly Cleaned** (Line 59, 187)
> 
> * `tempdir.keep()` prevents automatic cleanup
> * Line 187 cleanup with `let _ =` silently ignores errors
> * If test panics, directory leaks
> * **Fix:** Don't use `keep()`, let tempdir auto-cleanup on drop
> 
> **2. RwLock Usage Pattern** (Line 25)
> 
> * Production code uses `Arc<RwLock<HistoricalStorage>>` everywhere
> * Checkpoint expects bare `&HistoricalStorage` reference
> * Test acquires read lock which creates the race by design
> * **Clarify:** Is this intentionally demonstrating an artificial scenario, or should it match production patterns?
> 
> ## Moderate Issues
> **3. Non-deterministic Sleep** (Lines 77-80)
> 
> * 10ms sleep doesn't guarantee checkpoint acquires lock first
> * Makes test flaky on busy CI systems
> * **Fix:** Use proper synchronization (second Barrier) or document probabilistic nature
> 
> **4. Test Helper Naming** (Line 142)
> 
> * `__test_get_node_versions_iterator()` uses unconventional double underscore
> * **Fix:** Use `#[doc(hidden)]` or `#[cfg(test)]` instead
> 
> **5. Error Handling** (Lines 37, 41, 113, 119)
> 
> * Multiple `.expect()` calls that panic
> * **Improvement:** Consider using `?` operator with Result return type
> 
> ## Minor Issues
> * **Line 16:** Missing issue number (replace #XXX)
> * **Lines 31-43:** Consider reducing from 100 to 10 initial nodes for faster tests
> 
> ## Positive Aspects
> 1. Excellent documentation - race condition explanation is clear and educational
> 2. Comprehensive validation - checks both orphaned versions and missing history
> 3. Real-world simulation - updates both current and historical storage
> 4. Proper test isolation with #[ignore] attribute
> 
> ## Security & Performance
> * No security concerns (test code, bounded loops, no unsafe)
> * Performance adequate (200 nodes is lightweight)
> 
> ## Verdict
> **Conditional Approval** - Needs fixes for resource leak and RwLock clarification. Validation logic is solid and valuable.
> 
> **Code Review Checklist:**
> 
> * ✅ Tests cover edge cases
> * ✅ Follows CODING_STANDARDS.md
> * ⚠️ Error handling needs improvement
> * ⚠️ Documentation good, needs issue number
> * ❌ Resource cleanup must be fixed
> 
> Great work identifying this race condition!

@jules

**Comment by madmax983** [comment]
> ## Code Review - PR #602: Add validation for snapshot consistency
> ### Overall Assessment
> This is a **well-designed test** that successfully reproduces a real race condition in the checkpoint system. The code quality is high, with excellent documentation and clear intent. However, there are several areas for improvement.
> 
> ### ✅ Strengths
> 1. **Excellent Test Design**: The test clearly reproduces the "torn read" race condition where a checkpoint captures inconsistent state between current and historical storage.
> 2. **Comprehensive Validation**: The dual validation checks (orphaned versions + missing history) thoroughly verify snapshot consistency.
> 3. **Clear Documentation**: Comments explaining the race condition mechanism (lines 68-77) are outstanding - they make the subtle timing issue easy to understand.
> 4. **Proper Resource Management**: Using `tempdir()` in the main thread with RAII cleanup (line 31-32) is the correct approach.
> 5. **Better API Naming**: Renaming `__test_get_node_versions_iterator` to `iter_node_versions` improves code readability.
> 
> ### ⚠️ Issues to Address
> #### 1. **Code Style Violation: `.unwrap()` in Production-Adjacent Code** (Critical)
> **Location**: tests/snapshot_race_condition.rs:48, 124
> 
> ```rust
> // Line 48
> historical.write().unwrap()
> 
> // Line 124
> historical.write().unwrap()
> ```
> 
> **Issue**: Per [CODING_STANDARDS.md](https://github.com/madmax983/GallifreyDB/blob/trunk/docs/CODING_STANDARDS.md), `.unwrap()` should never be used without extremely good reason. Even in tests, `.expect()` is preferred for better diagnostics.
> 
> **Fix**:
> 
> ```rust
> historical
>     .write()
>     .expect("Lock poisoned: historical storage lock corrupted")
>     .add_node_version(node_id, version_id, temporal, label, props)
>     .expect("Failed to add initial node version");
> ```
> 
> #### 2. **Missing Issue Reference** (Medium Priority)
> **Location**: Line 20
> 
> ```rust
> #[ignore] // Will fail until fix is implemented (Issue #XXX - Race Condition)
> ```
> 
> The test references "Issue #XXX" but this should link to a real GitHub issue tracking the race condition fix.
> 
> **Action Required**: Create or reference the actual issue number.
> 
> #### 3. **Redundant Comment** (Minor)
> **Location**: Lines 93-96
> 
> ```rust
> // Small delay to ensure checkpoint thread has time to acquire the lock first
> // and create the "torn read" condition described above.
> // Intentional delay to increase probability of checkpoint thread
> // acquiring lock before writer, creating the "torn read" condition
> ```
> 
> These two comments say the same thing. Consolidate to:
> 
> ```rust
> // Intentional delay to increase probability of checkpoint thread
> // acquiring lock before writer, creating the "torn read" condition
> // described above (lines 68-77).
> ```
> 
> #### 4. **Test Helper API Visibility Concern** (Medium Priority)
> **Location**: src/storage/historical.rs:2008
> 
> The renamed `iter_node_versions()` is marked `#[doc(hidden)]` but no longer has the `__test_` prefix that signaled "internal use only". This could encourage inappropriate production usage.
> 
> **Options**:
> 
> 1. Keep the name but add a more prominent warning in the doc comment
> 2. Use `#[cfg(test)]` to make it test-only (preferred if only used in tests)
> 3. Consider if this should be promoted to a legitimate public API if it's useful for diagnostics
> 
> **Recommendation**: Since this is used across multiple test files, keeping it as `#[doc(hidden)] pub` is acceptable, but update the doc comment:
> 
> ```rust
> /// **TEST ONLY**: Iterate over all node versions.
> ///
> /// This method exposes internal implementation details and must ONLY be used
> /// in integration tests for validation purposes. Do not use in production code.
> ///
> /// # Warning
> /// This API is unstable and may change without notice.
> #[doc(hidden)]
> pub fn iter_node_versions(&self) -> impl Iterator<Item = &NodeVersion> {
>     self.node_versions.values()
> }
> ```
> 
> #### 5. **Potential Flakiness from Sleep-Based Timing** (Low Priority)
> **Location**: Line 97
> 
> ```rust
> std::thread::sleep(std::time::Duration::from_millis(10));
> ```
> 
> While the comment correctly notes this is intentional, this timing-based approach could be flaky on heavily loaded CI systems.
> 
> **Consideration**: The test is already `#[ignore]`d, so this is acceptable for now. If this becomes a problem when the bug is fixed and the test is enabled, consider using synchronization primitives (additional barriers) instead of sleep.
> 
> ### 🔒 Security Review
> **Status**: ✅ No security concerns
> 
> * Proper error propagation throughout
> * No unsafe code introduced
> * All IDs use validated constructors (NodeId::new())
> * Resource cleanup properly handled via RAII
> 
> ### 🎯 Test Coverage
> The validation logic (lines 165-188) provides excellent coverage:
> 
> * ✅ Detects orphaned versions (historical without current)
> * ✅ Detects missing history (current without historical)
> * ✅ Uses proper recovery flow to inspect checkpoint contents
> 
> ### 📊 Performance Considerations
> **Status**: ✅ Not applicable (test code)
> 
> The test intentionally creates contention to reproduce a bug. No performance concerns for test code.
> 
> ### ✅ Checklist Status
> * [x]  **Tests pass** (test is intentionally ignored until bug is fixed)
> * [x]  **Code formatted** (appears correctly formatted)
> * [ ]  **Clippy passes** - Need to verify, but likely issues with `.unwrap()` usage
> * [x]  **Coverage maintained** - Test code doesn't affect coverage thresholds
> * [x]  **Documentation updated** - Excellent inline documentation
> * [x]  **Temporal invariants preserved** - Test validates these
> * [x]  **No unsafe code** - All code is safe
> * [x]  **Strong typing** - Proper use of NodeId, VersionId, etc.
> 
> ### 🔧 Required Changes Before Merge
> 1. **Must Fix**:
>    
>    * Replace `.unwrap()` calls with `.expect()` (lines 48, 124)
>    * Add real issue number to `#[ignore]` annotation (line 20)
> 2. **Should Fix**:
>    
>    * Consolidate redundant comments (lines 93-96)
>    * Enhance doc comment on `iter_node_versions()` to emphasize test-only usage
> 3. **Nice to Have**:
>    
>    * Consider if the API rename warrants a broader discussion about test helper visibility
> 
> ### 📝 Summary
> This PR adds valuable test infrastructure to detect a real concurrency bug. The test is well-written with excellent documentation. After addressing the `.unwrap()` usage and adding the issue reference, this will be ready to merge.
> 
> **Recommendation**: ✅ **Approve with minor changes required**
> 
> The changes are low-risk (test-only code) and provide significant value by documenting a known race condition that needs to be fixed.

@jules

**Comment by madmax983** [comment]
@jules you are in danger of getting this PR closed. This ignored test has a placeholder issue. Either create an issue and put it in her . . . fix the text. or I will throw this code out. #[ignore] // Will fail until fix is implemented (Issue #XXX - Race Condition)

Tests are even passing in CI, and you have about 60% more test coverage to write. Please address these in the next commit, or this will be a waste of a task.
